libjcat/jcat-self-test.c

```diff
@@ -2613,9 +2613,308 @@ TestVerifyInclusionProof(void)
+				  gint64 snapshot,
+				  gint64 node_level,
+				  GByteArray **node)
+{
+	gint64 level = 0;
+	gint64 last_node = snapshot - 1;
+	g_autoptr(GByteArray) subtree_root = NULL;
+	if (snapshot == tree->leavesProcessed) {
+		if (node != NULL && tree->tree->len > node_level) {
+			GPtrArray *this_level = g_ptr_array_index(tree->tree, node_level);
+			if (*node != NULL)
+				g_byte_array_unref(*node);
+			if (node_level > 0) {
+				*node = g_byte_array_ref(
+				    g_ptr_array_index(this_level, this_level->len - 1));
+			} else {
+				*node = g_byte_array_ref(g_ptr_array_index(this_level, last_node));
+			}
+		}
+		return inmemoryTreeRoot(tree);
+	}
+
+	g_assert_cmpint(snapshot, <, tree->leavesProcessed);
+
+	while ((last_node & 1) == 1) {
+		if (node != NULL && node_level == level) {
+			GPtrArray *this_level = g_ptr_array_index(tree->tree, level);
+			if (*node != NULL)
+				g_byte_array_unref(*node);
+			*node = g_byte_array_ref(g_ptr_array_index(this_level, last_node));
+		}
+		last_node >>= 1;
+		level++;
+	}
+
+	subtree_root = g_byte_array_ref(
+	    g_ptr_array_index((GPtrArray *)g_ptr_array_index(tree->tree, level), last_node));
+
+	if (node != NULL && node_level == level) {
+		if (*node != NULL)
+			g_byte_array_unref(*node);
+		*node = g_byte_array_ref(subtree_root);
+	}
+
+	while (last_node != 0) {
+		if ((last_node & 1) == 1) {
+			GPtrArray *this_level = g_ptr_array_index(tree->tree, level);
+			GByteArray *new_subtree_root =
+			    jcat_rfc6962_hash_children(g_ptr_array_index(this_level, last_node - 1),
+						       subtree_root);
+			g_byte_array_unref(subtree_root);
+			subtree_root = new_subtree_root;
+		}
+
+		last_node >>= 1;
+		level++;
+		if (node != NULL && node_level == level) {
+			if (*node != NULL)
+				g_byte_array_unref(*node);
+			*node = g_byte_array_ref(subtree_root);
+		}
+	}
+	return subtree_root;
+}
+
+static GByteArray *
+inmemoryTreeRootAtSnapshot(inmemoryTree *tree, gint64 snapshot)
+{
+	if (snapshot == 0) {
+		return g_bytes_unref_to_array(sha256EmptyTreeHash());
+	}
+	if (snapshot > inmemoryTreeLeafCount(tree)) {
+		return NULL;
+	}
+	if (snapshot >= tree->leavesProcessed) {
+		return inmemoryTreeUpdateToSnapshot(tree, snapshot);
+	}
+	return inmemoryTreeRecomputePastSnapshot(tree, snapshot, 0, NULL);
+}
+
+static GByteArray *
+inmemoryTreeCurrentRoot(inmemoryTree *tree)
+{
+	return inmemoryTreeRootAtSnapshot(tree, inmemoryTreeLeafCount(tree));
+}
+
+static GPtrArray *
+inmemoryTreePathFromNodeToRootAtSnapshot(inmemoryTree *tree,
+					 gint64 node,
+					 gint64 level,
+					 gint64 snapshot)
+{
+	GPtrArray *path = g_ptr_array_new_with_free_func((GDestroyNotify)g_byte_array_unref);
+	gint64 last_node = (snapshot - 1) >> level;
+	if (snapshot == 0 || level >= tree->levelCount || node > last_node ||
+	    snapshot > inmemoryTreeLeafCount(tree))
+		return path;
+	if (snapshot > tree->leavesProcessed) {
+		inmemoryTreeUpdateToSnapshot(tree, snapshot);
+	}
+	while (last_node) {
```

Comment:
maybe add != 0 to make this clear it's an int.

Reply:
Done.